### ISSUE TITLE
 Possible problem with <symbol /> viewBox rendering

### ISSUE THREAD

**Issue by vladchuk** [zathras/jovial_svg#54]
I am not sure if it is a bug, but I perused all the readme's and documentation I could get my hands on and still failed to conclude this **should not** work:

```
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" class="card" face="AH" height="3.5in"
    preserveAspectRatio="none" viewBox="-106 -164.5 212 329" width="2.25in">
    <defs>
        <symbol id="VHA" viewBox="-500 -500 1000 1000" preserveAspectRatio="xMinYMid">
            <path d="M-270 460L-110 460M-200 450L0 -460L200 450M110 460L270 460M-120 130L120 130" stroke="#c80000"
                stroke-width="80" stroke-linecap="square" stroke-miterlimit="1.5" fill="none"></path>
        </symbol>
        <symbol id="SHA" viewBox="-600 -600 1200 1200" preserveAspectRatio="xMinYMid">
            <path
                d="M0 -300C0 -400 100 -500 200 -500C300 -500 400 -400 400 -250C400 0 0 400 0 500C0 400 -400 0 -400 -250C-400 -400 -300 -500 -200 -500C-100 -500 0 -400 -0 -300Z"
                fill="#c80000"></path>
        </symbol>
    </defs>
    <rect width="211" height="328" x="-105.5" y="-164" rx="12" ry="12" fill="white" stroke="black"></rect>
    <use xlink:href="#SHA" height="54" width="54" x="-27" y="-27"></use>
    <use xlink:href="#VHA" height="32" width="32" x="-100.4" y="-152.5"></use>
    <use xlink:href="#SHA" height="26.769" width="26.769" x="-97.784" y="-115.5"></use>
    <g transform="rotate(180)">
        <use xlink:href="#VHA" height="32" width="32" x="-100.4" y="-152.5"></use>
        <use xlink:href="#SHA" height="26.769" width="26.769" x="-97.784" y="-115.5"></use>
    </g>
</svg>
```
...but it only renders the black card outline, instead of Ace of Hearts.

Why I think it **should**? Well, it renders correctly in Firefox, Edge, Inkscape and a couple of VSCode plugins.
Best I can tell (based on my intuition) it has something to do with viewBox handling on <symbols/> and, according to the docs, those are supported. Since the file is very small, perhaps you can spot the issues right away, or at least provide some feedback. FWIW, it also fails with flutter_svg, but in a different way (only a red rectangle is produced).

Unrelated, but I would appreciate your input all the same:
1. On the surface, your package and futter_svg are very similar. Are they really? Sorry, I don't know enough to judge for myself :)
2. The map of Ukraine is very prominent in your demo. Are you in any way affiliated or do you just like the colors? :)

Thank you for reading and, hopefully, commenting!

**Comment by vladchuk** [comment]
Good to know I wasn't completely off in my assessment.

Yeah, poor choice of words regarding Ukraine, but you pretty much answered it. I originally started with flutter_svg, since I thought it was a bit easier to use, but it didn't work with my super-efficient deck (not created by me!), and I decided to give jovial a try. Efficiency is something I care about quite a bit as well, though.

Thank you for you prompt response.

**Comment by zathras** [comment]
Ah, yes -- the library does process a symbol's `viewbox` attribute, but _only_ if a width and height are specified.  I don't see a reason why it wouldn't be possible to calculate a width/height for the children of the symbol, which is probably what the browsers are doing.  SVG is littered with this kind of thing:  Parameters that are optional, and browsers that make heroic efforts to render something reasonable when they're missing.

Also, the library is ignoring the width/height specified on a `use` node, which probably isn't right?  I'll have to check the specs to see if that width/height is supposed to scale the children to fit, and if so, under what circumstances...  It's been a while since I read the spec closely for that sort of thing.

**Comment by vladchuk** [comment]
Alas, I know nothing about SVG, but here's my 2c, all the same. 

Indeed, specifying the width/height on symbols make them render according to the what is specified. However, I don't think this is correct, since the same symbol can be (and is, in my deck) used multiple times, with different sizes. According to the docs (https://developer.mozilla.org/en-US/docs/Web/SVG/Element/symbol), default values are "auto" - this is probably what is used by other rendering tools. As a matter of fact, those other tools seem to ignore this extra info and continue to render the image correctly.

By the same token, I think that <use/> width/height _should not be ignored_, for the reason I mention above (reuse in different contexts). FWIW, according to the same documentation, default value for those are "0"s. So this behavior would seem incorrect, indeed.

Thank you for looking into it!

**Comment by zathras** [comment]
Thanks for doing that testing!  That's about what I was planning to do, and it confirms what I thought was going on.

I bet that browsers aren't really _ignoring_ the width/height on the symbol, exactly.  What I think they're doing is scaling the symbol's viewbox to the width/height specified in the symbol tag.  But then, when the use tag uses the symbol, it scales that scaled image to a new width/height -- which in this case is identical, to the observed effect is nothing.  I bet if you removed the width/height from the use tag, you'd see the width/height of the symbol tag controlling the size.

So, I think the main thing I'm missing in the library is paying attention to the width/height in the `use` tag.

For completeness, I should look into the behavior when there's no viewbox, but a width/height is specified in either place.  My guess is that it calculates the "viewbox" by calculating a bounding box around the children of the symbol tag.

**Comment by vladchuk** [comment]
You are correct: removing the size from the use tag causes the size on the symbol to take over:) And removing the viewBox screws up rendering in a major way (in my case, blows the symbol up to some gigantic size).

**Comment by zathras** [comment]
Aha!  The interaction between `use` and `symbol` is specified at https://www.w3.org/TR/2011/REC-SVG11-20110816/single-page.html#struct-UseElement , section 5.6.  Turns out, `width` and `height` on a `use` tag are *only* meaningful if the referenced element is `symbol` or `svg` (and `jovial_svg` doesn't support embedded `svg` elements).  I'm pretty sure I implemented `use` before `symbol` ;-)

Anyhoo, that resolves one of the complexities I was thinking about.  `symbol` usually has an explicitly defined bounding box, and just about always has a well-defined bounding box (except maybe for a degenerate case, like one that's empty).  It's less clear if this is true or not for other nodes that can be referenced from `use`.
